README.md

```diff
@@ -1,6 +1,102 @@
+
+It also takes the exact call data for a call to `settle`.
+The flash-loan router is a solver for CoW Protocol and calls `settle` directly once the flash loans have been obtained.
+Only CoW-Protocol solvers can call this function.
+Solver authentication is done by interrogating the same contract that manages solver authentication for the settlement contract.
+
+Tokens and lenders are external contracts, while the router and each specific borrowers are dedicated contract implemented in this repository.
+
+The borrowers are the contracts that are called back by the lender once the flash loan is initiated; they are the contracts that receive the flash-loan proceeds and that are eventually responsible to repay the loan.
+
+The only way to move funds out of a borrower is through an ERC-20 approval for some spender.
+Approvals can be set by calling the `approve` function on the borrower from the context of a settlement.
+For safe operations, like an approval for the settlement contract to spend the funds of the borrower, it's enough to set the approval once for an unlimited amount and reuse the same approval in future settlements.
+
+At the start of the settlement, it's expected that the loaned funds are transferred from the borrowers to where they are needed. For example, this can be the settlement contract itself, or the address of a user who wants to use the loan to retrieve the collateral needed to avoid liquidations.
+In general, solvers have full flexibility in deciding how loaned funds are allocated.
+
+The settlement is also reponsible for repaying the flash loans.
+The specific repayment mechanism depends on the lender, but a common process is having the settlement contract send back the borrowed funds to the borrower and set an approval to the lender for spending the funds of the borrower: then the lender is going to pull back the funds with an ERC-20 `transferFrom` after the settlement is terminated.
+Inability to pay for a flash loan will most likely be met by a reverting transaction.
+
+We support the following flash-loan lenders:
+
+- Any lender that is compatible with [ERC-3156](https://eips.ethereum.org/EIPS/eip-3156) interface (for example [Maker](https://docs.makerdao.com/smart-contract-modules/flash-mint-module)).
+- [Aave](https://aave.com/docs/developers/flash-loans#overview).
+
+Support for further flash-loan lenders can be added in the future.
+
+### Example: a settlement with two loans, Aave and Maker
+
+We describe an example of how to execute a settlement after borrowing two flash loans (from Aave and Maker) through the flash-loan router.
+
+The following diagram is a simplified description of the expected calls in the execution process.
+Each line is a call from the originating contract to the target contract.
+Note that the call context is never given up in the diagram: each new call just increase the current call depth.
```

Comment:
I think you could make this very clear in the diagram itself by using [activations](https://mermaid.js.org/syntax/sequenceDiagram.html#activations)